File: 0155/two_stacks.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <numeric>

using namespace std;

struct Node {
  int val;
  Node* next;
  Node(int v) : val(v), next(nullptr) {};
};

class MinStack {
 public:
  /** initialize your data structure here. */
  MinStack() : top_(nullptr), min_(nullptr) {}

  ~MinStack() {
    Node* tmp;
    while (top_) {
      tmp = top_->next;
      delete top_;
      top_ = tmp;
      tmp = min_->next;
      delete min_;
      min_ = tmp;
    }
  }

  void push(int x) {
    Node* tmp;
    tmp = new Node(x);
    tmp->next = top_;
    top_ = tmp;
    if (!min_) {
      tmp = new Node(x);
    } else {
      int cur_min = getMin();
      tmp = (top() < cur_min) ? new Node(x) : new Node(cur_min);
    }
    tmp->next = min_;
    min_ = tmp;
  }

  void pop() {
    if (!top_) return;
    Node* tmp;
    tmp = top_;
    top_ = top_->next;
    delete tmp;
    tmp = min_;
    min_ = min_->next;
    delete tmp;
  }

  int top() {
    // RuntimeError when there is no element
    return top_->val;
  }

  int getMin() {
    // RuntimeError when there is no element
    return min_->val;
  }

 private:
  Node* top_;
  Node* min_;
};

int main() {
  MinStack* obj = new MinStack();
  obj->push(3);
  obj->push(-4);
  obj->push(1);
  assert(obj->top() == 1);
  assert(obj->getMin() == -4);
  obj->pop();
  assert(obj->top() == -4);
  assert(obj->getMin() == -4);
  obj->pop();
  assert(obj->top() == 3);
  assert(obj->getMin() == 3);
  delete obj;
}

File: 0155/two_stacks.py
class MinStack:
    def __init__(self):
        self._stack = []
        self._mins = []

    def push(self, val: int) -> None:
        self._stack.append(val)
        if len(self._mins) == 0:
            self._mins.append(val)
            return
        if val <= self.getMin():
            self._mins.append(val)

    def pop(self) -> None:
        assert len(self._stack) > 0
        val = self._stack.pop()
        if val == self._mins[-1]:
            self._mins.pop()

    def top(self) -> int:
        assert len(self._stack) > 0
        return self._stack[-1]

    def getMin(self) -> int:
        assert len(self._stack) > 0
        return self._mins[-1]


if __name__ == '__main__':
    min_stack = MinStack()
    min_stack.push(-2)
    min_stack.push(0)
    min_stack.push(-3)
    assert min_stack.getMin() == -3
    min_stack.pop()
    assert min_stack.top() == 0
    assert min_stack.getMin() == -2

    min_stack = MinStack()
    min_stack.push(0)
    min_stack.push(1)
    min_stack.push(0)
    assert min_stack.getMin() == 0
    min_stack.pop()
    assert min_stack.getMin() == 0

    min_stack = MinStack()
    min_stack.push(2147483646)
    min_stack.push(2147483646)
    min_stack.push(2147483647)
    assert min_stack.top() == 2147483647
    min_stack.pop()
    assert min_stack.getMin() == 2147483646
    min_stack.pop()
    assert min_stack.getMin() == 2147483646
    min_stack.pop()
    min_stack.push(2147483647)
    assert min_stack.top() == 2147483647
    assert min_stack.getMin() == 2147483647
    min_stack.push(-2147483648)
    assert min_stack.top() == -2147483648
    assert min_stack.getMin() == -2147483648
    min_stack.pop()
    assert min_stack.getMin() == 2147483647
